src/Catalyst.Core.Modules.Dfs/CoreApi/UnixFsApi.cs

```diff
@@ -0,0 +1,408 @@
+
+            return new UnixFsNode
+            {
+                Id = dag.Id,
+                Size = (long) dm.FileSize,
+                DagSize = dagSize + dag.Size,
+                Links = links
+            };
+        }
+
+        public async Task<IFileSystemNode> AddDirectoryAsync(string path,
+            bool recursive = true,
+            AddFileOptions options = default,
+            CancellationToken cancel = default)
+        {
+            options ??= new AddFileOptions();
+            options.Wrap = false;
+
+            // Add the files and sub-directories.
+            path = Path.GetFullPath(path);
+            var files = Directory.EnumerateFiles(path).OrderBy(s => s).Select(p => AddFileAsync(p, options, cancel));
+            if (recursive)
+            {
+                var folders = Directory.EnumerateDirectories(path).OrderBy(s => s)
+                   .Select(dir => AddDirectoryAsync(dir, true, options, cancel));
+                files = files.Union(folders);
+            }
+
+            var nodes = await Task.WhenAll(files).ConfigureAwait(false);
+
+            // Create the DAG with links to the created files and sub-directories
+            var links = nodes.Select(node => node.ToLink()).ToArray();
+            var fsn = await CreateDirectoryAsync(links, options, cancel).ConfigureAwait(false);
+            fsn.Name = Path.GetFileName(path);
+            return fsn;
+        }
+
+        private async Task<UnixFsNode> CreateDirectoryAsync(IEnumerable<IFileSystemLink> links,
+            AddFileOptions options,
+            CancellationToken cancel)
+        {
+            var dm = new DataMessage
+            {
+                Type = DataType.Directory
+            };
+            var pb = new MemoryStream();
+            Serializer.Serialize(pb, dm);
+            var fileSystemLinks = links as IFileSystemLink[] ?? links.ToArray();
+            var dag = new DagNode(pb.ToArray(), fileSystemLinks);
+
+            // Save it.
+            var cid = await GetBlockService(options).PutAsync(
+                data: dag.ToArray(),
+                multiHash: options.Hash,
+                encoding: options.Encoding,
+                pin: options.Pin,
+                cancel: cancel).ConfigureAwait(false);
+
+            return new UnixFsNode
+            {
+                Id = cid,
+                Links = fileSystemLinks,
+                IsDirectory = true
+            };
+        }
+
+        public async Task<IFileSystemNode> ListFileAsync(string path,
+            CancellationToken cancel = default)
+        {
+            var r = await _nameApi.ResolveAsync(path, true, false, cancel).ConfigureAwait(false);
+            var cid = Cid.Decode(r.Remove(0, 6));
+
+            var block = await _blockApi.GetAsync(cid, cancel).ConfigureAwait(false);
+
+            switch (cid.ContentType)
+            {
+                // TODO: A content-type registry should be used.
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)